### PR TITLE
Fail when using an untrusted builder and no lifecycle image available

### PR DESCRIPTION
**Pull request by natalieparellano**
Also consider lifecycle image supported when version is 0.6.1 (this version
of the lifecycle does not require the user to be valid in the image, so a
lifecycle image is possible).

Signed-off-by: Natalie Arellano <narellano@vmware.com>

### CODE REVIEW

**Comment by jromero** [comment]
@ekcasey / @sclevine Given that this is targeting a release that is **Feature Complete**. Can I get at least one additional approval to add it to this release?

**Comment by natalieparellano** [comment]
User acceptance for lifecycle 0.6.1:

### Untrusted:
```
$  ./out/pack build localhost:5000/test-app -p ~/workspace/samples/apps/java-maven/ -B test-0.6.1 --publish --network host
0.6.1: Pulling from buildpacksio/lifecycle
Digest: sha256:81cf254bf41f56fd1f527622724f760f7cf44d6714e5355526d7e008b548b0b4
Status: Image is up to date for buildpacksio/lifecycle:0.6.1
===> DETECTING
...
Successfully built image localhost:5000/test-app
```
(note the `0.6.1: Pulling from buildpacksio/lifecycle`)

### Trusted:
```
$  ./out/pack build localhost:5000/test-app -p ~/workspace/samples/apps/java-maven/ -B test-0.6.1 --publish --network host --trust-builder
===> DETECTING
...
Successfully built image localhost:5000/test-app
```
(note there is NO `0.6.1: Pulling from buildpacksio/lifecycle`)

### Untrusted but not publishing:
```
$  ./out/pack build localhost:5000/test-app -p ~/workspace/samples/apps/java-maven/ -B test-0.6.1
alpine: Pulling from cnbs/sample-stack-run
4167d3e14976: Already exists 
3b5ba8762a59: Already exists 
0defd2e00834: Already exists 
Digest: sha256:ff5952abb5476176ddd68790c1ba69ef0114f701c0565acc94ad067100571449
Status: Downloaded newer image for cnbs/sample-stack-run:alpine
===> DETECTING
...
Successfully built image localhost:5000/test-app
```
(note there is NO `0.6.1: Pulling from buildpacksio/lifecycle`)

**Comment by natalieparellano** [comment]
@jromero I see that codecov is not happy - just want to note that I did TDD this one out (by changing the lifecycle version in the units from 0.3.0 -> 0.6.1 I got a failure).

File: build.go [hunk 1]
```diff
@@ -162,21 +162,20 @@ func (c *Client) Build(ctx context.Context, opts BuildOptions) error {
...
 		lifecycleOpts.UseCreator = true
+		// no need to fetch a lifecycle image, it won't be used
 		return c.lifecycle.Execute(ctx, lifecycleOpts)
 	}
 
-	lifecycleImageSupported := !lifecycleVersion.LessThan(semver.MustParse("0.7.5"))
-	if !lifecycleImageSupported {
-		c.logger.Warnf("Lifecycle %s does not have an associated lifecycle image.", lifecycleVersion.String())
-	} else {
+	lifecycleImageSupported := lifecycleVersion.Equal(builder.VersionMustParse("0.6.1")) || !lifecycleVersion.LessThan(semver.MustParse("0.7.5"))
```
Comment (natalieparellano) [inline build.go:176]: Do we want to add 0.6.0?

**Comment by natalieparellano** [comment]
Deleted the previous user acceptance as it's no longer relevant.

The latest commits **change the behavior** a bit based on recent conversations.

When a builder is not trusted, we will ALWAYS try to use a "lifecycle image" for analyze, restore, and export (previously, only when `--publish`). Rationale: there are still some security concerns around mounting the docker daemon socket with untrusted builders.

If no such lifecycle image exists, we will FAIL. This is the case for lifecycle versions 0.7.0-0.7.4 (no such image is possible, as the lifecycle requires the CNB user to be created), and pre-0.6.1. The user can run `pack trust-builder` or pass `--trust-builder` to continue with their workflow.

**Comment by natalieparellano** [comment]
0.6.1

Untrusted:
```
$  ./out/pack build test-app -p ~/workspace/samples/apps/java-maven/ -B test-0.6.1
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:ff5952abb5476176ddd68790c1ba69ef0114f701c0565acc94ad067100571449
Status: Image is up to date for cnbs/sample-stack-run:alpine
0.6.1: Pulling from buildpacksio/lifecycle
Digest: sha256:81cf254bf41f56fd1f527622724f760f7cf44d6714e5355526d7e008b548b0b4
Status: Image is up to date for buildpacksio/lifecycle:0.6.1
===> DETECTING
[detector] samples/java-maven 0.0.1
===> ANALYZING
[analyzer] Warning: Image "index.docker.io/library/test-app:latest" not found
===> RESTORING
===> BUILDING
[builder] ---> Java buildpack
...
```
(note: `0.6.1: Pulling from buildpacksio/lifecycle`)

Trusted:
```
$  ./out/pack build test-app -p ~/workspace/samples/apps/java-maven/ -B test-0.6.1 --trust-builder
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:ff5952abb5476176ddd68790c1ba69ef0114f701c0565acc94ad067100571449
Status: Image is up to date for cnbs/sample-stack-run:alpine
===> DETECTING
[detector] samples/java-maven 0.0.1
===> ANALYZING
[analyzer] Restoring metadata for "samples/java-maven:jdk" from app image
[analyzer] Restoring metadata for "samples/java-maven:maven_m2" from cache
===> RESTORING
[restorer] Restoring data for "samples/java-maven:jdk" from cache
[restorer] Restoring data for "samples/java-maven:maven_m2" from cache
===> BUILDING
[builder] ---> Java buildpack
...
```

File: build.go (hunk 1 above)
Comment (jromero) [reply build.go:176]: Based on the latest state of known builders and the age of that release, I would think it's unlikely that 0.6.0 is used heavily. A minor lifecycle bump to said builder would be minimal effort for added security.

**Comment by natalieparellano** [comment]
0.7.4:

Untrusted:
```
$  ./out/pack build test-app -p ~/workspace/samples/apps/java-maven/ -B test-0.7.4
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:ff5952abb5476176ddd68790c1ba69ef0114f701c0565acc94ad067100571449
Status: Image is up to date for cnbs/sample-stack-run:alpine
ERROR: Lifecycle 0.7.4 does not have an associated lifecycle image. Builder must be trusted.
```

Trusted:
```
$  ./out/pack build test-app -p ~/workspace/samples/apps/java-maven/ -B test-0.7.4 --trust-builder
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:ff5952abb5476176ddd68790c1ba69ef0114f701c0565acc94ad067100571449
Status: Image is up to date for cnbs/sample-stack-run:alpine
===> CREATING
[creator] ---> DETECTING
[creator] samples/java-maven 0.0.1
[creator] ---> ANALYZING
[creator] Restoring metadata for "samples/java-maven:jdk" from app image
[creator] Restoring metadata for "samples/java-maven:maven_m2" from cache
[creator] ---> RESTORING
[creator] Restoring data for "samples/java-maven:jdk" from cache
[creator] Restoring data for "samples/java-maven:maven_m2" from cache
[creator] ---> BUILDING
...
```

**Comment by natalieparellano** [comment]
0.7.5:

Untrusted:
```
$  ./out/pack build test-app -p ~/workspace/samples/apps/java-maven/ -B test-0.7.5
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:ff5952abb5476176ddd68790c1ba69ef0114f701c0565acc94ad067100571449
Status: Image is up to date for cnbs/sample-stack-run:alpine
0.7.5: Pulling from buildpacksio/lifecycle
Digest: sha256:246e6c365f53306583286bcb307b6dd7700e66188d4e9aabf1e4d3139af0f8e1
Status: Image is up to date for buildpacksio/lifecycle:0.7.5
===> DETECTING
[detector] samples/java-maven 0.0.1
===> ANALYZING
[analyzer] Restoring metadata for "samples/java-maven:jdk" from app image
[analyzer] Restoring metadata for "samples/java-maven:maven_m2" from cache
===> RESTORING
[restorer] Restoring data for "samples/java-maven:jdk" from cache
[restorer] Restoring data for "samples/java-maven:maven_m2" from cache
===> BUILDING
...
```
(note: `0.7.5: Pulling from buildpacksio/lifecycle`)

Trusted:
```
$  ./out/pack build test-app -p ~/workspace/samples/apps/java-maven/ -B test-0.7.5 --trust-builder
alpine: Pulling from cnbs/sample-stack-run
Digest: sha256:ff5952abb5476176ddd68790c1ba69ef0114f701c0565acc94ad067100571449
Status: Image is up to date for cnbs/sample-stack-run:alpine
===> CREATING
...
```

**Pull request merged**